### PR TITLE
Feature/orca 381 Build internal_reconcile_report_mismatch

### PR DESCRIPTION
**Pull request by AndrewDorn**
## Summary of Changes

Please write a sentence or two summarizing the proposed change.

Addresses [ORCA-381: Build internal_reconcile_report_mismatch](https://bugs.earthdata.nasa.gov/browse/ORCA-381)

## Changes

* Added lambda for reporting mismatches
* Added `comment` for describing race conditions.

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

## How Has This Been Tested?

Deployed to and run in AWS.

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the necessary documentation (remove those that do not apply)
    - [x] CHANGELOG.md
    - [x] User documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
    - [x] Unit tests
- [x] New and existing unit tests pass locally with my changes
    - [x] Automated tests passing (if not fork)
- [x] Any dependent changes have been merged and published in downstream modules
- [x] My code has passed security scanning
    - [x] Snyk
    - [x] git-secrets

### CODE REVIEW

File: tasks/internal_reconcile_report_mismatch/schemas/output.json [hunk 1]
```diff
@@ -1,91 +1,95 @@
...
-        "discrepancyType": {
-          "type": "string",
-          "description": "ype of discrepancy found during reconciliation."
-      }
+          },
+          "discrepancyType": {
+            "type": "string",
+            "description": "Type of discrepancy found during reconciliation."
+          },
+          "comment": {
```
Comment (AndrewDorn) [inline tasks/internal_reconcile_report_mismatch/schemas/output.json:89]: Open to alternate approaches/expansions to this. Currently can only be 'Error may be due to race condition, and should be checked manually.'

File: website/docs/developer/research/research-lambda-container.md [hunk 2]
```diff
@@ -327,8 +327,10 @@ Recommend use of the above architecture to resolve concerns with timeouts when h
...
 This script can be built into a Docker container and deployed to the result of [ECR reserach](https://bugs.earthdata.nasa.gov/browse/ORCA-375). https://bugs.earthdata.nasa.gov/browse/ORCA-429
 This new script can then be deployed as a task definition in AWS. https://bugs.earthdata.nasa.gov/browse/ORCA-432
-The task definition can be run periodically to empty the queue of any internal reconciliation jobs. https://bugs.earthdata.nasa.gov/browse/ORCA-430
-We can either remove triggering from post_to_queue_and_trigger_step_function or update it to trigger the Fargate task when called. https://bugs.earthdata.nasa.gov/browse/ORCA-431
+We should also create our own ECS cluster to decouple from Cumulus. https://bugs.earthdata.nasa.gov/browse/ORCA-433
+The task definition can be run periodically in the ECS cluster to empty the queue of any internal reconciliation jobs. https://bugs.earthdata.nasa.gov/browse/ORCA-430
+We can either remove triggering from post_to_queue_and_trigger_step_function or update it to trigger the Fargate task when called.
+In either case, the old workflow can be removed. https://bugs.earthdata.nasa.gov/browse/ORCA-431
```
Comment (AndrewDorn) [inline website/docs/developer/research/research-lambda-container.md:334]: Neglected to save before committing for my last PR. Whoops.

File: tasks/internal_reconcile_report_mismatch/schemas/output.json [hunk 3]
```diff
@@ -1,91 +1,95 @@
...
+          "keyPath",
+          "cumulusArchiveLocation",
+          "orcaEtag",
+          "s3Etag",
+          "orcaLastUpdate",
+          "s3LastUpdate",
+          "orcaSizeInBytes",
+          "s3SizeInBytes",
+          "discrepancyType",
+          "comment"
```
Comment (rizbihassan) [inline tasks/internal_reconcile_report_mismatch/schemas/output.json:38]: some of these might become optional after https://bugs.earthdata.nasa.gov/browse/ORCA-412

**Review by bhazuka — APPROVED** [review]
Overall looks good. Added some comments about some items to think about/contemplate.

File: tasks/internal_reconcile_report_mismatch/internal_reconcile_report_mismatch.py [hunk 4]
```diff
@@ -0,0 +1,262 @@
...
+                    "page_index": page_index,
+                    "page_size": PAGE_SIZE,
+                }
+            ],
+        )
+
+        mismatches = []
+        for sql_result in sql_results:
+            mismatches.append(
+                {
```
[the remark this answers is not in the document]
Comment (AndrewDorn) [reply tasks/internal_reconcile_report_mismatch/internal_reconcile_report_mismatch.py:110]: Yup, that's present.

File: modules/lambdas/main.tf [hunk 5]
```diff
@@ -118,6 +118,35 @@ resource "aws_lambda_function" "perform_orca_reconcile" {
   }
 }
 
+# internal_reconcile_report_mismatch - Receives job id and page index from end user and returns reporting information of files that have records in the S3 bucket but are missing from ORCA catalog.
+# ==============================================================================
+resource "aws_lambda_function" "internal_reconcile_report_mismatch" {
+  ## REQUIRED
+  function_name = "${var.prefix}_internal_reconcile_report_mismatch"
+  role          = var.restore_object_role_arn
```
[the remark this answers is not in the document]
Comment (AndrewDorn) [reply modules/lambdas/main.tf:126]: https://bugs.earthdata.nasa.gov/browse/ORCA-440

[the remark this answers is not in the document]
Comment (AndrewDorn) [reply modules/lambdas/main.tf:126]: Once we have an agreed upon structure, I'll start doing this for new code. As-is, I'm avoiding having two people developing coding standards.

File: modules/lambdas/main.tf [hunk 6]
```diff
@@ -118,6 +118,35 @@ resource "aws_lambda_function" "perform_orca_reconcile" {
...
+
+  ## OPTIONAL
+  description      = "Receives job id and page index from end user and returns reporting information of files that have records in the S3 bucket and the ORCA catalog, but the records disagree."
+  filename         = "${path.module}/../../tasks/internal_reconcile_report_mismatch/internal_reconcile_report_mismatch.zip"
+  handler          = "internal_reconcile_report_mismatch.handler"
+  memory_size      = var.orca_reconciliation_lambda_memory_size
```
[the remark this answers is not in the document]
Comment (AndrewDorn) [reply modules/lambdas/main.tf:132]: I've tossed a couple comments on cards like https://bugs.earthdata.nasa.gov/browse/ORCA-387 and https://bugs.earthdata.nasa.gov/browse/ORCA-410 as we should have a larger discussion about these values... Values like orca_ingest_lambda_memory_size warp thinking in a way I don't really like, and I feel that we should rearchitect the entire structure.

**Pull request merged**